File: apps/api/plane/package_flow/models/execution.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

"""Runner, claim and run models (FR-G03..FR-G07, INV-01, INV-04, INV-09).

Customer code never executes in the Plane web/API/live process; these rows
only coordinate an external runner that connects outbound.
"""

from django.conf import settings
from django.db import models
from django.db.models import Q

from .base import ExtensionBaseModel, IssueScopedModel


class RunnerProfile(ExtensionBaseModel):
    """A registered external runner (local, customer-hosted or managed; O01)."""

    class Kind(models.TextChoices):
        LOCAL = "local", "Developer machine"
        CUSTOMER = "customer", "Customer-hosted runner"
        MANAGED = "managed", "Managed runner"

    name = models.CharField(max_length=255)
    kind = models.CharField(max_length=16, choices=Kind.choices, default=Kind.LOCAL)
    # Responsible human Plane user; the runner acts as an *agent* principal.
    owner = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name="+")
    token_hash = models.CharField(max_length=64, unique=True)
    token_prefix = models.CharField(max_length=12)
    is_active = models.BooleanField(default=True)
    last_seen_at = models.DateTimeField(null=True, blank=True)
    policy = models.JSONField(default=dict, blank=True)

    class Meta:
        db_table = "pf_runner_profiles"


class Claim(IssueScopedModel):
    """Atomic exclusive (or explicitly collaborative) claim with lease + fencing (FR-G04)."""

    class Status(models.TextChoices):
        ACTIVE = "active", "Active"
        RELEASED = "released", "Released"
        EXPIRED = "expired", "Expired"
        REVOKED = "revoked", "Revoked"

    repository_binding = models.ForeignKey(
        "package_flow.RepositoryBinding", on_delete=models.CASCADE, null=True, blank=True, related_name="+"
    )
    # Unit key: issue + binding (or "none" for non-repo packages) — used for the exclusive constraint.
    unit_key = models.CharField(max_length=128)
    exclusive = models.BooleanField(default=True)
    holder = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name="+")
    runner = models.ForeignKey(RunnerProfile, on_delete=models.CASCADE, null=True, blank=True, related_name="claims")
    approval = models.ForeignKey(
        "package_flow.ExecutionApproval", on_delete=models.PROTECT, null=True, blank=True, related_name="claims"
    )
    fencing_token = models.BigIntegerField()
    lease_expires_at = models.DateTimeField()
    last_heartbeat_at = models.DateTimeField(null=True, blank=True)
    status = models.CharField(max_length=16, choices=Status.choices, default=Status.ACTIVE)
    released_at = models.DateTimeField(null=True, blank=True)

    class Meta:
        db_table = "pf_claims"
        constraints = [
            # At most one active exclusive claim per unit — enforced by the database (AC05).
            models.UniqueConstraint(
                fields=["unit_key"],
                condition=Q(status="active", exclusive=True, deleted_at__isnull=True),
                name="pf_claim_one_active_exclusive",
            )
        ]


class FencingCounter(models.Model):
    """Monotonic fencing sequence per claim unit."""

    unit_key = models.CharField(max_length=128, primary_key=True)
    value = models.BigIntegerField(default=0)

    class Meta:
        db_table = "pf_fencing_counters"


class ExecutionRun(IssueScopedModel):
    """One controlled execution bound to exact revision, approval and claim (INV-01/04)."""

    class Status(models.TextChoices):
        QUEUED = "queued", "Queued"
        CLAIMED = "claimed", "Claimed"
        RUNNING = "running", "Running"
        WAITING = "waiting", "Waiting"
        FAILED = "failed", "Failed"
        CANCELLED = "cancelled", "Cancelled"
        FINISHED = "finished", "Finished"

    revision = models.ForeignKey("package_flow.PackageRevision", on_delete=models.PROTECT, related_name="runs")
    approval = models.ForeignKey("package_flow.ExecutionApproval", on_delete=models.PROTECT, related_name="runs")
    claim = models.ForeignKey(Claim, on_delete=models.PROTECT, related_name="runs")
    runner = models.ForeignKey(RunnerProfile, on_delete=models.PROTECT, null=True, blank=True, related_name="runs")
    responsible = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.PROTECT, related_name="+")
    status = models.CharField(max_length=16, choices=Status.choices, default=Status.QUEUED)
    mode = models.CharField(max_length=16, default="human")  # human | agent
    agent_adapter = models.CharField(max_length=64, blank=True, default="")
    base_commits = models.JSONField(default=dict, blank=True)
    limits = models.JSONField(default=dict, blank=True)
    spend_minor = models.BigIntegerField(default=0)
    run_token_hash = models.CharField(max_length=64, blank=True, default="")
    run_token_expires_at = models.DateTimeField(null=True, blank=True)
    started_at = models.DateTimeField(null=True, blank=True)
    finished_at = models.DateTimeField(null=True, blank=True)
    last_heartbeat_at = models.DateTimeField(null=True, blank=True)
    progress = models.JSONField(default=dict, blank=True)
    result = models.JSONField(default=dict, blank=True)
    cancel_requested_at = models.DateTimeField(null=True, blank=True)
    pause_reason = models.CharField(max_length=64, blank=True, default="")

    class Meta:
        db_table = "pf_execution_runs"
        ordering = ("-created_at",)


class RunAction(IssueScopedModel):
    """Controlled runner action log (push, commit, check …), fenced (AC06, FR-G06)."""

    run = models.ForeignKey(ExecutionRun, on_delete=models.CASCADE, related_name="actions")
    action = models.CharField(max_length=32)
    fencing_token = models.BigIntegerField()
    accepted = models.BooleanField(default=False)
    reason = models.CharField(max_length=255, blank=True, default="")
    detail = models.JSONField(default=dict, blank=True)

    class Meta:
        db_table = "pf_run_actions"
        ordering = ("created_at",)

File: packages/types/src/project-hub/execution.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

/** Runner, claims, runs (API §3). */

export type TRunnerKind = "local" | "customer" | "managed";

export type TRunnerProfile = {
  id: string;
  name: string;
  kind: TRunnerKind;
  owner_id: string;
  token_prefix: string;
  is_active: boolean;
  last_seen_at: string | null;
  created_at: string | null;
};

/** `POST W/runners/` → token is returned exactly once. */
export type TRunnerRegistration = {
  id: string;
  token: string;
  name?: string;
  kind?: TRunnerKind;
  token_prefix?: string;
};

export type TClaimStatus = "active" | "released" | "expired" | "revoked";

/** `claim_data` (+ web projection fields from `GET P/work-items/{id}/claims`). */
export type TPackageClaim = {
  id: string;
  work_item_id: string;
  repository_binding_id: string | null;
  exclusive: boolean;
  holder_id: string;
  runner_id: string | null;
  approval_id: string | null;
  fencing_token: number;
  lease_expires_at: string;
  status: TClaimStatus;
  last_heartbeat_at?: string | null;
  released_at?: string | null;
  created_at?: string | null;
  runner_name?: string | null;
  repository_name?: string | null;
};

export type TPackageClaimList = { results: TPackageClaim[] };

export type TClaimCreate = {
  approval_id?: string;
  repository_binding_id?: string;
  exclusive?: boolean;
  lease_seconds?: number;
};
export type TClaimCreated = TPackageClaim;

export type TRunStatus = "queued" | "claimed" | "running" | "waiting" | "failed" | "cancelled" | "finished";
export type TRunMode = "human" | "agent";

export type TPackageRun = {
  id: string;
  work_item_id: string;
  revision_id: string;
  approval_id: string;
  claim_id: string;
  runner_id: string | null;
  responsible_id: string;
  status: TRunStatus;
  mode: TRunMode;
  agent_adapter: string;
  base_commits: Record<string, string>;
  limits: Record<string, unknown>;
  spend_minor: number;
  started_at: string | null;
  finished_at: string | null;
  last_heartbeat_at: string | null;
  progress: { summary?: string; percent?: number; question?: string; [key: string]: unknown };
  result: Record<string, unknown>;
  cancel_requested_at: string | null;
  pause_reason: string;
  manifest_hash: string;
};

/** `GET P/work-items/{id}/runs` → `{results: [Run]}`. */
export type TPackageRunList = { results: TPackageRun[] };

export type TRunCreate = { approval_id: string; claim_id: string; mode: TRunMode; agent_adapter?: string };
